### DISCUSSION TITLE
Elsa 2.4.0 🍭

### DISCUSSION THREAD

**Discussion by sfmskywalker** [elsa-workflows/elsa-core#2635] [Design & tech]
## BREAKING CHANGES ⚠️

The **WorkflowInstances** table received a new column: `WorkflowDefinitionVersionId`.
This is a required field, which means that running the updated DB migrations will fail when run while the table has existing records.

If you do not wish to first delete the contents of this table, make sure to create a backup first (which you should do at all times before applying an updated migration).

The value of the `WorkflowDefinitionVersionId` needs to be taken from the **WorkflowDefinitions** table's `Id` field.

## New Features ✨
- Retention settings (auto-delete completed workflow instances)
- Run workflow from designer (Requires new `Elsa.WorkflowTesting.Api` package and update Startup with `services.AddWorkflowTestingServices()` and `app.MapWorkflowTestHub` - see `Elsa.Samples.Workflow` for an example)
- Configure automatic layout direction. Default is top to bottom, but can now be configured from the UI to left to right, right to left and bottom to top (#2619)
- FileSystemWatcher activities
- Configure HTTP Endpoint activity with authentication settings (require authenticated requests and ASP.NET Core policy requirements)
- Configure Send HTTP Request activity with desired response content formatter (plus the ability to add your own by implementing `IHttpResponseContentReader`)
- Added RabbitMQ activities (#2547)

## Improvements 🍏
- Added designer legend with quick hints about certain designer operations such as how to delete connections and change connection between activities.
- Performance improvement (noticeable when there are many workflow definitions in the system)
- Trim Azure Service Bus queue name at 50 characters (ASB returns error when queue name limit of 50 characters is exceeded)
- Ability to select multiple workflow instances from multiple pages.
- Add variables tab to instance view (#2442)
- Add base64 encoding JS functions
- Add `getActivityId` JS function
- Increase dashboard extensibility. All menu items and routes are now customizable by plugins
- Disable buggy copy & paste. To be re-implemented (possibly as part of Elsa 3 designer back-port)
- Add support for bitflag list items (#1654)
- Add intellisense support for dynamic HttpEndpoint output (#2526)
- Update appveyor to push debug symbols
- Use strongly-typed arrays for typescript definitions (#2584)

## Fixes 🛠️
- Publish correct event (WorkflowInstanceCancelled) (#1560)
- Fix YesSQL provider for bookmarks (#1604)
- Fix for oracle persistence provider - in order to use data more than 2000 char we have to use NCLOB ,in oracle we have to explicitly say the column is NCLOB otherwise it would be considered Nvarchar(2000) (#2450) 
- Dynamic Activity Outcomes from activity property  descriptor (#2473)
- fixed semaphore locking for azureservicebus activities (#2494)
- Fix ClearTimer activity (#2429)
- Replace Dictionary with ConcurrentDictionary (#2503)
- Update workflow launchpad to prevent duplicate correlated workflow instances
- Prevent activities with [Browsable(false)] from being included (#2411)
- Accessing Output via Liquid (#2509)
- Various JSON serialization fixes and improvements
- Many other bug fixes

This release was made possible thanks to generous contributions from members from the community.

A Big Thank You to everyone who raised issues and contributed amazing features & fixes! This release is awesome because of all of you🙏🏻 

<hr /><em>This discussion was created from the release <a href='https://github.com/elsa-workflows/elsa-core/releases/tag/2.4.0.1'>Elsa 2.4.0 🍭</a>.</em>

**Comment by pentower** [comment]
It seems like some activities stop working properly. No Intellisense for Run JavaScript activity. HTTP Response activity now sends its response body only in raw text format.

**Reply by sfmskywalker** [reply]
I tried to reproduce the two issues you mentioned, but it seems to work fine for me. Any chance you can provide me with steps to reproduce or a sample project?
